qutip/core/blochredfield.py

```diff
@@ -1,20 +1,49 @@
 import os
 import inspect
 import numpy as np
-from qutip.settings import settings as qset
+from typing import overload
 
+from qutip.settings import settings as qset
 from . import Qobj, QobjEvo, liouvillian, coefficient, sprepost
 from ._brtools import SpectraCoefficient, _EigenBasisTransform
 from .cy.coefficient import InterCoefficient, Coefficient
 from ._brtensor import _BlochRedfieldElement
-
+from ..typing import CoeffProtocol
 
 __all__ = ['bloch_redfield_tensor', 'brterm']
 
 
-def bloch_redfield_tensor(H, a_ops, c_ops=[], sec_cutoff=0.1,
-                          fock_basis=False, sparse_eigensolver=False,
-                          br_dtype='sparse'):
+@overload
+def bloch_redfield_tensor(
+    H: Qobj,
```

Comment:
It's unclear to me how to read these overloads and to tell what they allow.

The three cases chosen seem to be a subset of the full cartesian product of possible types and their expected return type (please don't add all possible cases).

In the first overload, it's unclear to me how the return can be a `Qobj` if the `a_ops` or `c_ops` are time-dependent.

Passing `Qobj` for `H` or `c_ops` seems to be an allowed type for all three overloads, but the allowed return types are different in each case.

The last two overloads seem to be identical except that the third allows a return of `Qobj | QobjEvo` making the second redundant.

Probably `@overload` is good in some cases of intermediate complexity, but probably not a good fit for cases like this where the output type depends on the types of three other parameters.

Reply:
Returning multiple types `Qobj | QobjEvo` is causing more issues than helping. Static type checker will cause error if you try to use the result any way that is not supported by all return types. In this case we there is `@overload` for complex case and `TypeVar` for simple ones. 

In the first case, `c_ops` accepting `QobjEvo` is an error. Thank you for catching. The coefficient in `a_ops` is for the spectrum function only and cannot be time dependent if the operator is a `Qobj`.

The second case is any other accepted case.

The last case not an overload and is not used by type checker. I could remove types there if it remove confusion.

Reply:
Hmm. I don't have a great suggestion for how to resolve this.

When we started with type hints for QuTiP we wanted to create simple hints that would help our users understand what kinds of objects could be passed and help us be consistent in what kinds of objects we supported.

These complex overload annotations are not that.

Since I don't have a great suggestion, let me review the rest and let you decide how to proceed here.